dali/python/nvidia/dali/plugin/pytorch/__init__.py

```diff
@@ -795,3 +853,266 @@ def __next__(self) -> List[Dict[str, torch.Tensor]]:
+    execute in the main process via the GPU.
+
+    Background: As the PyTorch workers run on separate processes, using the GPU directly from each
+    of those is not ideal for performance, due to the usage of several CUDA contexts.
+    """
+
+    def __init__(self, input_names):
+        """
+        Initializes a new DALI proxy instance.
+
+        Args:
+            input_names (list): list of strings representing the inputs to the pipeline. Those should match
+                                the names of the ``external_source`` nodes in the DALI pipeline.
+        """
+        self.input_names = input_names
+        self.num_inputs = len(input_names)
+        # Multi-process queue used to transfer data from the pytorch workers to the main process
+        self.send_q = mp.Queue()
+        # Multi-process queue used by the main process to remember the actual order of execution of the requests
+        self.order_q = mp.Queue()
+        # Torch worker id, to be filled on first call to worker_id()
+        self._worker_id = None
+        # Iteration index for the current worker
+        self.data_idx = 0
+
+    def worker_id(self):
+        if self._worker_id is None:
+            self._worker_id = torch.utils.data.get_worker_info().id
+        return self._worker_id
+
+    class PipelineOutputRef:
+        """
+        Placeholder for a pipeline output reference, after the iteration has been scheduled to DALI.
+        """
+        def __init__(self, info):
+            self.info = info
+
+    def schedule_batch(self, inputs):
+        """
+        Schedule a pipeline run to DALI, by queuing the worker id, the iteration index and the inputs
+        """
+        # Identifier of this request
+        info = (self.worker_id(), self.data_idx)
+        with nvtx.annotate(f"dali_proxy.send_q.put {info}", color="blue"):
+            self.send_q.put((info, inputs))
+        self.data_idx = self.data_idx + 1
+        # Returns a placeholder, which is replaced with the actual data once the iteration completes
+        return DALIProxy.PipelineOutputRef(info)
+
+    class PipelineRunRef:
+        """
+        Placeholder for a pipeline run reference, which is returned by the data worker instead of the actual data
+
+        The PyTorch worker returns this trivial object, only containing information about this proxy instance and
+        the input data to the pipeline. Later in the collate function, we send the data for execution to DALI.
+        """
+        def __init__(self, dali_proxy, inputs):
+            self.dali_proxy = dali_proxy
+            self.inputs = inputs
+            assert len(self.inputs) == dali_proxy.num_inputs
+
+    def transform(self, *inputs):
+        """
+        The 'transform' function consists of returning a reference to the pipeline run
+        """
+        assert len(inputs) == self.num_inputs, f"Unexpected number of inputs: {len(inputs)}"
```

Comment:
This error might be reported to the user if they do something wrong, right? In this case this should be a proper exception